File: HackerRank/cracking the interview/shortestreach.py
class Node:							#represents a node of a graph
	def __init__(self):
		self.visited=False			#flag variable to avoid loops by revisiting nodes.
		self.neighbors=[]
		self.distance=-1			#distance from root. -1 means disconnected.

def printShortestPaths(nodelist,root):	#implements BFS by level order traversal of graph
	nodelist[root].visited=True
	nodelist[root].distance=0
	levelqueue=[root]
	while len(levelqueue)>0:
		curr=levelqueue[0]
		for i in nodelist[curr].neighbors:
			if not nodelist[i].visited:
				levelqueue.append(i)
				nodelist[i].visited=True
				nodelist[i].distance=nodelist[curr].distance+6
		levelqueue.pop(0)

	for i in nodelist:
		if i.distance!=0:
			print(i.distance,end=' ')
	print()

q=int(input())
for _ in range(q):
	n,m=[int(x) for x in input().split()]
	nodelist=[]
	for _ in range(n):
		newnode=Node()
		nodelist.append(newnode)
	for _ in range(m):
		a,b=[int(x) for x in input().split()]
		nodelist[a-1].neighbors.append(b-1)
		nodelist[b-1].neighbors.append(a-1)
	root=int(input())-1
	printShortestPaths(nodelist,root)

File: HackerRank/cracking the interview/shortestreach.cpp
#include<bits/stdc++.h>
using namespace std;

class Node
{
public:
    bool visited;
    vector<int> adj;
    int distance;
    Node()
    {
        visited=false;
        distance=-1;
    }
};

void printShortestPaths(vector<Node> nodelist, int root)
{
    nodelist[root].visited=true;
    nodelist[root].distance=0;
    queue<int> levelqueue;
    levelqueue.push(root);
    while (!levelqueue.empty())
    {
        int curr=levelqueue.front();
        for (int i=0;i<nodelist[curr].adj.size();i++)
            if (!(nodelist[nodelist[curr].adj[i]].visited))
                {
                    levelqueue.push(nodelist[curr].adj[i]);
                    nodelist[nodelist[curr].adj[i]].visited=true;
                    nodelist[nodelist[curr].adj[i]].distance=nodelist[curr].distance+6;
                }
        levelqueue.pop();
    }
    for (int i=0;i<nodelist.size();i++)
        if (nodelist[i].distance)
            cout<<nodelist[i].distance<<" ";
    cout<<endl;
}

int main()
{
	int q;
	cin>>q;
	while(q--)
	{
		int n,m;
		cin>>n>>m;
		vector<Node> nodelist;
		while(n--)
        {
            Node newnode;
            nodelist.push_back(newnode);
        }
        while(m--)
        {
            int u,v;
            cin>>u>>v;
            nodelist[u-1].adj.push_back(v-1);
            nodelist[v-1].adj.push_back(u-1);
        }
        int root;
        cin>>root;
        root--;
        printShortestPaths(nodelist,root);

	}
}
